Support CsWinRT 3.0 multi-targeting and WindowsSdkPackageMinimumRevision

dsplaisted opened the pull request:
This is a continuation of #49721

This PR allows multi-targeting CsWinRT 3.0, by leveraging the build number of the target Windows SDK.
That is, e.g.:
- `net10.0-windows10.0.22621.0` ---> CsWinRT 2.x
- `net10.0-windows10.0.22621.1` ---> CsWinRT 3.0

In the implementation, rather than using separate ref packs for CSWinRT 2 and 3, which the previous PR did, this PR uses profiles of the Microsoft.Windows.SDK.NET.Ref framework reference to differentiate between the versions of CSWinRT.

This also adds support for a WindowsSdkPackageMinimumRevision property.

Sergio0694 committed:
Support '1.' TFM for CsWinRT 3.0

Sergio0694 committed:
Add 'UsingPreviewCsWinRT3_0Feature' warning

Sergio0694 committed:
Add test for new preview warning

Sergio0694 committed:
Run 'msbuild /t:UpdateXlf'

Sergio0694 committed:
Update 'TargetPlatformVersion' parsing

Sergio0694 committed:
Update targeting pack selection logic

Sergio0694 committed:
Update '_WarnForCsWinRT3_0PreviewFeatureEnabled'

Sergio0694 committed:
Define 'CSWINRT3_0' for CsWinRT 3.0

Sergio0694 committed:
Add unit tests

Sergio0694 committed:
Revert incorrect 'Strings.resx' change

Sergio0694 committed:
Update .xlf files

Sergio0694 committed:
Fix parsing normalized SDK version

Sergio0694 committed:
Fix framework references creation for '.1' TFM

Sergio0694 committed:
Add unit tests for normalized defines

dsplaisted committed:
Merge branch 'main' into user/sergiopedri/cswinrt3-multi-targeting

dsplaisted committed:
Switch to using profiles in the same framework reference to differentiate between CSWinRT 2 and 3

dsplaisted committed:
Add support for WindowsSdkPackageMinimumRevision

manodasanW committed:
Fix included profiles

dsplaisted committed:
Add CSWinRT 2 and 3 multi-targeting test

dsplaisted committed:
Don't normalize to .0 TargetPlatformVersion revision for CSWinRT 3

manodasanW committed:
Fix profile names

manodasanW committed:
Normalize implicit define

Sergio0694 commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildAWindowsDesktopProject.cs:
```diff
@@ -478,6 +478,130 @@ public void ItDoesNotWarnAnymoreWhenBuildingAProjectWithUseUwpProperty()
+                .Should()
+                .Pass()
+                .And
+                .NotHaveStdOutContaining("NETSDK1229");
```
Mmh isn't this wrong? This test _should_ emit the warning, right? 🤔

dsplaisted commented on src/Layout/redist/targets/GenerateBundledVersions.targets:
```diff
@@ -1317,7 +1317,20 @@ Copyright (c) .NET Foundation. All rights reserved.
                               />
 
-    <!-- Supported Windows versions -->
+    <!--
+      Supported Windows versions using CsWinRT 3.0 (.NET 10 is the baseline). This leverages the revision of
+      the Windows SDK version to allow multi-targeting, which is required to allow customers to gradually migrate.
+      Both CsWinRT 3.0 and 2.x will be supported for a period of time, and packages can target both of them.
+    -->
+    <WindowsSdkSupportedTargetPlatformVersion Include="10.0.26100.1" WindowsSdkPackageVersion="$(MicrosoftWindowsSDKNETRef10_0_26100PackageVersion)" MinimumNETVersion="10.0" />
+    <WindowsSdkSupportedTargetPlatformVersion Include="10.0.22621.1" WindowsSdkPackageVersion="$(MicrosoftWindowsSDKNETRef10_0_22621PackageVersion)" MinimumNETVersion="10.0" />
+    <WindowsSdkSupportedTargetPlatformVersion Include="10.0.22000.1" WindowsSdkPackageVersion="$(MicrosoftWindowsSDKNETRef10_0_22000PackageVersion)" MinimumNETVersion="10.0" />
+    <WindowsSdkSupportedTargetPlatformVersion Include="10.0.20348.1" WindowsSdkPackageVersion="$(MicrosoftWindowsSDKNETRef10_0_20348PackageVersion)" MinimumNETVersion="10.0" />
+    <WindowsSdkSupportedTargetPlatformVersion Include="10.0.19041.1" WindowsSdkPackageVersion="$(MicrosoftWindowsSDKNETRef10_0_19041PackageVersion)" MinimumNETVersion="10.0" />
+    <WindowsSdkSupportedTargetPlatformVersion Include="10.0.18362.1" WindowsSdkPackageVersion="$(MicrosoftWindowsSDKNETRef10_0_18362PackageVersion)" MinimumNETVersion="10.0" />
+    <WindowsSdkSupportedTargetPlatformVersion Include="10.0.17763.1" WindowsSdkPackageVersion="$(MicrosoftWindowsSDKNETRef10_0_17763PackageVersion)" MinimumNETVersion="10.0" />
+
+    <!-- Supported Windows versions using CsWinRT 2.x -->
```
I put separate items here because these are linked to the TargetPlatformVersion, so we need items both for the .0 and .1 versions.  The actual package versions are the same between the corresponding items.

It would be possible to do this differently, but this seemed to be the simplest to me.

Sergio0694 replied:
Ah, gotcha, yeah makes perfect sense! Thank you for the additional context. Perhaps we could add a line to that comment at line 1321 to also mention this, for future reference? In case anyone else wonders the same I mean.

manodasanW committed:
Fix SupportedOSPlatformVersion and tests

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets:
```diff
@@ -65,6 +65,17 @@ Copyright (c) .NET Foundation. All rights reserved.
+    -->
+    <_TargetPlatformVersionUsesCsWinRT3>false</_TargetPlatformVersionUsesCsWinRT3>
+    <_TargetPlatformVersionUsesCsWinRT3 Condition="'$(TargetPlatformIdentifier)' == 'Windows' and '$(TargetPlatformVersion)' != '' and $([System.Version]::Parse('$(TargetPlatformVersion)').Revision) == 1">true</_TargetPlatformVersionUsesCsWinRT3>
+    <EffectiveTargetPlatformVersion Condition="'$(TargetPlatformIdentifier)' == 'Windows' and '$(TargetPlatformVersion)' != ''">$([System.Version]::Parse('$(TargetPlatformVersion)').Major).$([System.Version]::Parse('$(TargetPlatformVersion)').Minor).$([System.Version]::Parse('$(TargetPlatformVersion)').Build).0</EffectiveTargetPlatformVersion>
```
Should we set `EffectiveTargetPlatformVersion` in all cases, even if it's the same as `TargetPlatformVersion`?  Then in places that consume it, we could always use the `EffectiveTargetPlatformVersion` property, instead of having to condition on whether it's set.

manodasanW committed:
Fix tests and always set EffectiveTargetPlatformVersion

manodasanW committed:
Fix TargetPlatformVersion being set in Directory.Build.targets

manodasanW commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets:
```diff
@@ -234,7 +243,15 @@ Copyright (c) .NET Foundation. All rights reserved.
+        This allows to drop the .1 at the end of the version in CsWinRT 3.0 scenarios and 
+        use a common define for both CsWinRT 3.0 and CsWinRT 2.x.
+      -->
+      <_SupportedPlatformCompatibleVersions Include="@(SdkSupportedTargetPlatformVersion)"
+                                            Condition=" %(Identity) != '' and '%(SdkSupportedTargetPlatformVersion.NormalizedSupportedTargetPlatformVersion)' == '' and $([MSBuild]::VersionLessThanOrEquals(%(Identity), $(TargetPlatformVersion))) " />
+      <_SupportedPlatformCompatibleVersions Include="@(SdkSupportedTargetPlatformVersion->'%(NormalizedSupportedTargetPlatformVersion)')"
+                                            Condition=" '%(SdkSupportedTargetPlatformVersion.NormalizedSupportedTargetPlatformVersion)' != '' and $([MSBuild]::VersionLessThanOrEquals('%(SdkSupportedTargetPlatformVersion.NormalizedSupportedTargetPlatformVersion)', $(TargetPlatformVersion))) " />
```
When `_SupportedPlatformCompatibleVersions` is used right below, it is used as `@(_SupportedPlatformCompatibleVersions->Distinct()` so that removes the duplicates.

manodasanW commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -1497,9 +1497,14 @@ Copyright (c) .NET Foundation. All rights reserved.
     <TargetPlatformDisplayName>$([Microsoft.Build.Utilities.ToolLocationHelper]::GetPlatformSDKDisplayName($(TargetPlatformIdentifier), $(TargetPlatformVersion)))</TargetPlatformDisplayName>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <!-- If the Windows targets wasn't called to default it, then just default it to the TargetPlatformVersion. -->
+    <EffectiveTargetPlatformVersion Condition="'$(EffectiveTargetPlatformVersion)' == ''">$(TargetPlatformVersion)</EffectiveTargetPlatformVersion>
+  </PropertyGroup>
```
I initially made `Microsoft.NET.TargetFrameworkInference.targets` always default it, but then I ran into issues with `Directory.Build.Targets` can set `WindowsPlatformVersion` and that was after the `TargetFrameworkInference.targets`, so I had to handle that case in `Microsoft.Net.Windows.targets` which is based on `EffectiveTargetPlatformVersion` not being set.  Given the use of `Directory.Build.targets` which probably can also be used to set the platform, I put this here to catch such scenarios.

manodasanW commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.WindowsSdkSupportedTargetPlatforms.props:
```diff
@@ -22,7 +22,12 @@ Copyright (c) .NET Foundation. All rights reserved.
+        <!-- Normalize the last part of TFM to .0 when ending with .1 as that indicates the use of the CsWinRT 3.0 verison. -->
+        <_NormalizedWindowsSdkSupportedTargetPlatformVersion Include="@(WindowsSdkSupportedTargetPlatformVersion)">
+            <NormalizedSupportedTargetPlatformVersion 
+                Condition="$([System.Text.RegularExpressions.Regex]::IsMatch('%(Identity)', '^\d+\.\d+\.\d+\.1$'))">$([System.Version]::Parse('%(Identity)').Major).$([System.Version]::Parse('%(Identity)').Minor).$([System.Version]::Parse('%(Identity)').Build).0</NormalizedSupportedTargetPlatformVersion>
```
We can probably use Version.Parse here, no specific reason for the regex.

manodasanW replied:
Addressed.

manodasanW committed:
Apply suggestion from @dsplaisted

Co-authored-by: Daniel Plaisted <dsplaisted@gmail.com>

manodasanW committed:
Fix typo

manodasanW committed:
Use version parse instead

dsplaisted committed:
Add tests and additional error case for WindowsSdkPackageMinimumRevision

dsplaisted committed:
Collect binlog for test

dsplaisted committed:
Murge upstream changes

dsplaisted committed:
Copilot implementation of new logic for stage 2 bundled versions

Copilot prompt #1:

Replace the #class:'Microsoft.DotNet.Build.Tasks.OverrideAndCreateBundledNETCoreAppPackageVersion':1164-10242 Task.

The new version should take as parameters the path to the stage 0  bundled versions file as well as the version from stage 2.  It should copy some version numbers from items in the stage 0 version to the stage 2 version, and then save the updated stage 2 version.

It should update items where the TargetFramework metadata is not the latest target framework supported.  There are multiple item types that need to be updated, and they need to be matched on specific metadata.

KnownFrameworkReference: match on Include, TargetFramework metadata.  Update LatestRuntimeFrameworkVersion and TargetingPackVersion
KnownAppHostPack: match on Include, TargetFramework metadata.  Update AppHostPackVersion
KnownCrossgen2Pack: match on Include, TargetFramework.  Update Crossgen2PackVersion
KnownILCompilerPack: match on Include, TargetFramework.  Update ILCompilerPackVersion
KnownRuntimePack: match on Include, TargetFramework, RuntimePackLabels.  Update LatestRuntimeFrameworkVersion
KnownILLinkPack: match on Include, TargetFramework.  Update ILLinkPackVersion.

If there is more than one match based on the specified metadata, generate an error.

For the metadata which is not used to match and is not updated, if the value differs between the stage 0 and stage 2 versions, log a message.

Copilot prompt #2:

Don't look at just the first ItemGroup in each file, load items from all ItemGroups found.

When looking for the latest target framework, use the NuGetFramework class to parse the TargetFramework values, and to compare them.

dsplaisted committed:
Fixup/edits to copilot code

dsplaisted commented:
@marcpopMSFT I've pushed changes to this PR that significantly change OverrideAndCreateBundledNETCoreAppPackageVersion, which you enabled in #50418.

The problem with the previous logic is that it basically uses the stage 0 bundled versions for stage 2 with some version numbers updated.  This was necessary when the real bundled versions was in the separate installer repo.  Now that the repos are merged though, we generate the fully correct bundled versions file in the SDK.  If we replace that with the stage 0 version, then when we make changes to what's in the bundled versions file, they won't be available for tests until the updates flow back through stage 0.

So what the logic now does is preserve most of what is in the newly built bundled versions file, and only copy over version numbers for downlevel runtimes from stage 0.  I think that should solve the issues.

manodasanW commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildAWindowsDesktopProject.cs:
```diff
@@ -478,6 +478,139 @@ public void ItDoesNotWarnAnymoreWhenBuildingAProjectWithUseUwpProperty()
+                {
+                    ["Program.cs"] = """
+                    #if CSWINRT3_0
+                    #error CSWINRT3_0 is defined
```
This test is more if you are building a project with .0, make sure that none of the CSWINRT_3 things are defined.

dsplaisted committed:
Remove unnecessary usings
